crates/ruff_python_semantic/src/model.rs

```diff
@@ -966,6 +966,69 @@ impl<'a> SemanticModel<'a> {
     pub const fn future_annotations(&self) -> bool {
         self.flags.contains(SemanticModelFlags::FUTURE_ANNOTATIONS)
     }
+
+    /// Return an iterator over all bindings shadowed by the given [`BindingId`], within the
+    /// containing scope, and across scopes.
+    pub fn shadowed_bindings(
+        &self,
+        scope_id: ScopeId,
+        binding_id: BindingId,
+    ) -> impl Iterator<Item = ShadowedBinding> + '_ {
+        let mut first = true;
+        let mut binding_id = binding_id;
+        std::iter::from_fn(move || {
+            // First, check whether this binding is shadowing another binding in a different scope.
+            if std::mem::take(&mut first) {
+                if let Some(shadowed_id) = self.shadowed_bindings.get(&binding_id).copied() {
+                    return Some(ShadowedBinding {
+                        binding_id,
+                        shadowed_id,
+                        same_scope: false,
+                    });
+                }
+            }
+
+            // Otherwise, check whether this binding is shadowing another binding in the same scope.
+            if let Some(shadowed_id) = self.scopes[scope_id].shadowed_binding(binding_id) {
```

Comment:
Unrelated to this PR. Why do we need a global shadowed_bindings and a scoipe specfici shadowed binding tracking?

Reply:
Unfortunately, Flake8 has different rules around _when_ it flags for shadowed-while-unused depending on whether the shadowing is within-scope or across scopes.

Namely, when flagging shadowed-while-unused across scopes, it only flags when the shadowed variable is an import like:

```python
import os

def f():
  os = 1
```

But not:

```python
os = 1

def f():
  os = 1
```

However, when flagging shadowed-while-unused within scopes, it does flag:

```python
def f():
  os = 1
  os = 1
```